### ISSUE TITLE
dotnet repl auto-references mscorlib which breaks on CoreCLR

### ISSUE THREAD

**Issue by anurse** [dotnet/cli#320]
This may be a Roslyn issue, but I'm starting it off here because that's where I repro'd it.

Do the following:

1. Run `csi` (`dotnet-repl-csi` is broken, I'm fixing that now)
2. Type `System.ConsoleColor.Red`
  1. :+1: `(1,1): error CS0234: The type or namespace name 'ConsoleColor' does not exist in the namespace 'System' (are you missing an assembly reference?)` (this is **expected**)
3. Type `#r "System.Console.dll"`
4. Type `System.ConsoleColor.Red`
  1. :+1: `Red` (again, **expected**, System.Console.dll was properly loaded)
5. Type `System.Console.WriteLine`
  1. :boom: `(1,1): error CS0433: The type 'Console' exists in both 'System.Console, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e'`

It looks like `csi` is automatically referencing `mscorlib.dll` causing these type conflicts. AFAIK, CoreCLR-based apps/scripts shouldn't be referencing `mscorlib`. As far as I can tell, this is coming from https://github.com/dotnet/roslyn/blob/master/src/Scripting/Core/Script.cs#L186

Feel free to move over to dotnet/roslyn, I wanted to file it here to make it clear that the `csi` in the dotnet-cli is the one that is broken (the desktop one seems to work fine because referencing mscorlib is OK there).

I'm trying to resolve #270 by converting build scripts to `csx` so they are cross-plat, but unfortunately this is blocking me :(

**Comment by piotrpMSFT** [comment]
Jared, @AArnott, can you guys help get the right person looking at this one?

**Comment by jaredpar** [comment]
Moved over to the scripting team (CC @dotnet/roslyn-interactive) 

I'm not familiar with that code but it seems odd to unconditionally reference mscorlib.  Naively I would assume that it is tied to the `/nostdlib` option same as the rest of the compiler.

**Comment by anurse** [comment]
I'd certainly like to try and get this resolved ASAP. As far as I can tell, it makes it very difficult (maybe impossible...) to use CoreFX libraries in the CoreCLR-based `csi`.

**Comment by amcasey** [comment]
http://index/#Microsoft.CodeAnalysis.Scripting/Script.cs,195

**Comment by amcasey** [comment]
Maybe you have to use binaries that were themselves compiled on CoreCLR?  Or maybe that's just wrong.  @tmat would be the one to ask, but he's still out.

**Comment by jaredpar** [comment]
@amcasey these should be the CoreCLR CLI binaries

https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.Tools.Repl.Csi/project.json#L23

**Comment by anurse** [comment]
Yeah, that line is wrong in a CoreCLR context. CoreCLR compilations should never reference `mscorlib`, they should only reference the contract assemblies (`System.Runtime`, etc.)

**Comment by amcasey** [comment]
This is the IL I'm seeing for that line:

```
IL_0022:  ldtoken    [System.Runtime]System.Object
IL_0027:  call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
```

That seems correct, doesn't it?

**Comment by anurse** [comment]
Sure, but we're still seeing the issue. My interpretation of the cause may very well be off.

**Comment by amcasey** [comment]
@anurse, sorry, I meant that as a question rather than a challenge.  The IL looks sane to me, but when I step over that line in the debugger, `corlib` is set to `C:\Enlistments\roslyn-internal_master\Open\Binaries\Debug\core-clr\mscorlib.ni.dll`, which surprises me (and matches your theory of why you're seeing ambiguity).

**Comment by anurse** [comment]
No worries :).

I think the problem is the distinction between reference and implementation assemblies. At runtime, `typeof(object)` appears to return the forwarded `mscorlib` version of `System.Object`. I'd assume that's something that happens inside `GetTypeFromHandle` or elsewhere in the runtime. We should, instead, add a reference to `System.Runtime` directly rather than relying on `typeof(object)` to find it.

**Comment by amcasey** [comment]
I'm guessing the current implementation was an attempt to make it work with both mscorlib and System.Runtime.  I also remember @tmat saying that we specifically wanted to use the runtime assemblies everywhere, though I can't recall exactly why.  Maybe @cston has more context?

**Comment by anurse** [comment]
Using runtime assemblies everywhere in a scripting context makes some sense, but the problem is that users need to know how to resolve that distinction and select the right assemblies for their `#r` directives, etc.

**Comment by amcasey** [comment]
This looks relevant: https://github.com/dotnet/roslyn/issues/6094

I'm pretty sure @tmat has changes in flight in this area, so I'm a little reluctant to touch it.  Is there something we can do to unblock you until he returns?

**Comment by anurse** [comment]
If `csi` could accept `/nostdlib` and completely remove all references (except those provided by `/r` in the `csi` command line), that would be fairly low-impact and sufficient for now. In the context I'm working on (dotnet-cli build scripts), I have easy full access to the reference assemblies in a reliable location.

**Comment by tmat** [comment]
See proposal #5540

**Comment by tmat** [comment]
The issue has been addressed by dotnet/roslyn#8040

**Comment by tannergooding** [comment]
Should be fixed with #1209, if it is not fixed already.
